migration-engine/connectors/sql-migration-connector/src/flavour/mssql.rs

```diff
@@ -159,9 +161,54 @@ impl SqlFlavour for MssqlFlavour {
 
     async fn sql_schema_from_migration_history(
         &self,
-        _: &[MigrationDirectory],
-        _: &Connection,
+        migrations: &[MigrationDirectory],
+        connection: &Connection,
     ) -> ConnectorResult<SqlSchema> {
-        todo!("Needs the connection string crate, so leaving it unimplemented for now")
+        let database_name = format!("prisma_migrations_shadow_database_{}", uuid::Uuid::new_v4());
+        let create_database = format!("CREATE DATABASE [{}]", database_name);
+
+        connection.raw_cmd(&create_database).await?;
+
+        let mut jdbc_string: JdbcString = self.0.connection_string().parse().unwrap();
+        jdbc_string
+            .properties_mut()
+            .insert("database".into(), database_name.clone());
+        let temporary_database_url = jdbc_string.to_string();
+
+        tracing::debug!("Connecting to temporary database at {}", temporary_database_url);
+
+        let sql_schema = {
+            let temporary_database = crate::connect(&temporary_database_url).await?;
+            let create_schema = format!("CREATE SCHEMA [{schema}]", schema = self.schema_name());
+
+            temporary_database.raw_cmd(&create_schema).await?;
+
+            for migration in migrations {
+                let script = migration.read_migration_script()?;
+
+                tracing::debug!(
+                    "Applying migration `{}` to temporary database.",
+                    migration.migration_name()
+                );
+
+                temporary_database
+                    .raw_cmd(&script)
+                    .await
+                    .map_err(ConnectorError::from)
+                    .map_err(|connector_error| {
+                        connector_error.into_migration_does_not_apply_cleanly(migration.migration_name().to_owned())
+                    })?;
+            }
+
+            // the connection to the temporary database is dropped at the end of
+            // the block.
+            self.describe_schema(&temporary_database).await?
+        };
+
+        let drop_database = format!("DROP DATABASE [{}]", database = database_name);
```

Comment:
We made the shadow database  cleanup more robust on postgres and mysql in https://github.com/prisma/prisma-engines/pull/1417 — it would be great if we can align MSSQL on that (drop the db instead of early return on errors).

Reply:
Fixed in https://github.com/prisma/prisma-engines/pull/1372/commits/32128d5f6fc562707d2006f3fd95663b33d95df9